src/test/util/setup_common.cpp

```diff
@@ -134,21 +135,43 @@ BasicTestingSetup::BasicTestingSetup(const ChainType chainType, const std::vecto
         // By default, the data directory has a random name
         const std::string testdatadir{g_insecure_rand_ctx_temp_path.rand256().ToString().c_str()};
         m_path_root = fs::temp_directory_path() / "test_common_" PACKAGE_NAME / testdatadir;
+        fs::create_directories(m_path_root);
     } else {
         // Custom data directory
+        m_has_custom_datadir = true;
         const std::string testdatadir{m_node.args->GetArg("-testdatadir").value()};
         m_path_root = fs::path(testdatadir.c_str()) / "test_temp";
+
+        // Try to obtain the lock; if unsuccessful don't disturb the existing test.
+        auto lock_datadir = [&]() -> void {
+            try {
+                fs::create_directories(m_path_root);
+            } catch (const std::exception&) {
+            }
+            if (!fs::exists(m_path_root)) {
+                std::cerr << "Cannot create test data directory " + fs::PathToString(m_path_root) + '\n';
+                exit(1);
+            }
+            if (util::LockDirectory(m_path_root, ".lock", /*probe_only=*/false) != util::LockResult::Success) {
+                std::cerr << "Cannot obtain a lock on test data directory " + fs::PathToString(m_path_root) + '\n' +
+                                 "The test executable is probably already running.\n";
+                exit(1);
+            }
+        };
+        lock_datadir();
+
+        // Always start with a fresh data directory (must release the lock before remove on some platforms).
+        ReleaseDirectoryLocks();
         fs::remove_all(m_path_root);
+        lock_datadir();
```

Comment:
If a concurrent process locks the datadir right after the lock release, this process would remove the root path even when the other process is using it?

Reply:
Good catch, I was actually aware of this but concluded that it is such a small timing window, that it's not a problem in practice. The problem is that on Windows, you can't remove an open file (and having a file lock is a form of having the file open). I could make this conditional on `WIN32` (I think it is), but I'm not sure that's even worth it.